smarts/core/utils/math.py

```diff
@@ -372,6 +382,20 @@ def distance_point_to_polygon(point, polygon, perpendicular=False):
         return -1
 
 
+def bisect(lst: list, value: float, key_func=None) -> int:
+    if key_func is None:
+        key_func = lambda x: x
+    left = 0
+    right = len(lst)
+    while left < right:
+        mid = (left + right) // 2
+        if key_func(lst[mid]) <= value:
+            left = mid + 1
+        else:
+            right = mid
+    return left - 1
```

Comment:
It's cool that you wrote this yourself, but I meant the `bisect` module from the python standard library.   https://docs.python.org/3/library/bisect.html

 (Sorry, I should have been clearer!)

Even though your implementation seems fine, it's probably better to use the standard one, just in case they've optimized it somehow.

Reply:
I tried using the `bisect` module initially, but realized the key parameter does not work for the module. So I can try writing a `KeyWrapper` to bypass the issue if it is essential to use the module instead.

Reply:
Done!